letsencrypt/scripts/main.py

```diff
@@ -9,61 +9,100 @@
-                             "changes.")
-    parser.add_argument("-r", "--redirect", dest="redirect",
-                        action="store_const", const=True,
-                        help="Automatically redirect all HTTP traffic to HTTPS "
-                             "for the newly authenticated vhost.")
-    parser.add_argument("-n", "--no-redirect", dest="redirect",
-                        action="store_const", const=False,
-                        help="Skip the HTTPS redirect question, allowing both "
-                             "HTTP and HTTPS.")
-    parser.add_argument("-e", "--agree-tos", dest="eula", action="store_true",
-                        help="Skip the end user license agreement screen.")
-    parser.add_argument("-t", "--text", dest="use_curses", action="store_false",
-                        help="Use the text output instead of the curses UI.")
-    parser.add_argument("--test", dest="test", action="store_true",
-                        help="Run in test mode.")
+    add = parser.add_argument
+
+    add("-d", "--domains", metavar="DOMAIN", nargs="+")
+    add("-s", "--acme-server", "--server", default="letsencrypt-demo.org:443",
+        help="CA hostname (and optionally :port). The server certificate must "
+             "be trusted in order to avoid further modifications to the "
+             "client.")
+
+    add("-p", "--privkey", type=read_file,
+        help="Path to the private key file for certificate generation.")
+    add("-B", "--rsa-key-size", "--keysize", type=int, default=2048,
+        metavar="N", help="RSA key shall be sized N bits.")
+
+    add("-k", "--revoke", action="store_true", help="Revoke a certificate.")
+    add("-b", "--rollback", type=int, default=0, metavar="N",
+        help="Revert configuration N number of checkpoints.")
+    add("-v", "--view-config-changes", action="store_true",
+        help="View checkpoints and associated configuration changes.")
+    add("-r", "--redirect", action="store_true",
+        help="Automatically redirect all HTTP traffic to HTTPS for the newly "
+             "authenticated vhost.")
+
+    add("-e", "--agree-tos", dest="eula", action="store_true",
+        help="Skip the end user license agreement screen.")
+    add("-t", "--text", dest="use_curses", action="store_false",
+        help="Use the text output instead of the curses UI.")
+    add("--test", action="store_true", help="Run in test mode.")
+
+    # TODO: trailing slashes might be important! check and remove
+    add("--config-dir", default="/etc/letsencrypt/",
+        help="Configuration directory.")
+    add("--work-dir", default="/var/lib/letsencrypt/",
+        help="Working directory.")
+    add("--backup-dir", default="/var/lib/letsencrypt/backups/",
+        help="Configuration backups directory.")
+    add("--temp-checkpoint-dir",
+        default="/var/lib/letsencrypt/temp_checkpoint/",
+        help="Temporary checkpoint directory.")
+    add("--in-progress-dir",
```

Comment:
Their are quite a few related directories here... these used to be constructed using WORK_DIR and CONFIG_DIR.  I think it makes the job of operating a safe execution environment much more difficult and it will make testing harder.  Will we throw errors when intermediate directories are no longer created.  Certainly the code right now wasn't designed with so much independence in mind.  This also seems more difficult to port now that everything is completely specified separately.  You can't just change work-dir and config-dir and have everything work.  You must go through every variable and find where it should go in the new fs.

config-dir, working-dir and backup-dir, cert-dir, key-dir all seem reasonable to specify.  (Though the code will have to be checked to make sure that backup-dir isn't relying on work-dir too heavily)

rev-tokens-dir, cert-key-backup, temp-checkpoint-dir and in-progess-dir were never really meant to see the light of day and I think it will only confuse the user by giving them more options than they need.

I feel like specifying the names of certificates would be helpful.... but, it would be really nice to just specify the filename in cert-path.  It is way to easy to get cert-path and cert-dir messed up and then be completely confused when the program crashes.